Use client.set_many in HashClient.

martinnj opened the pull request:
This should fix #319 .

Some unit tests had to be changed since the data returned from `client.set_many` is different from when we iterate over `client.set` multiple times.

- `test_ignore_exec_set_many` - With the old approach if an exception occurred we could mark the individual key as failed and keep going. With `set_many` we only get a list of failed keys if no exception was thrown. Instead if `ignore_exc` is set, I chose to mark all keys as set.
- `test_set_many_unix` - It relies on the ordering of keys in a dict, which is unordered. I changed it from checking if a specific key failed, to just check a single key failed.

I don't have all the python interpreters installed on my machine, but tests pass for the following envs in tox:

- `py27`
- `py36`
- `py37`
- `py27-flake8`
- `py39-flake8`

`py39` tests failed because it was compiled without sqlite.

If I missed something or there is anything else that needs doing, please let me know.

jparise commented:
> `test_set_many_unix` - It relies on the ordering of keys in a dict, which is unordered. I changed it from checking if a specific key failed, to just check a single key failed.

I think this would be a good place to use [`OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict). Otherwise, the test isn't particularly strong because it doesn't verify that the `NOT_STORED` item is the one that's actually returned.

martinnj commented:
> 
> 
> > `test_set_many_unix` - It relies on the ordering of keys in a dict, which is unordered. I changed it from checking if a specific key failed, to just check a single key failed.
> 
> I think this would be a good place to use [`OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict). Otherwise, the test isn't particularly strong because it doesn't verify that the `NOT_STORED` item is the one that's actually returned.

I took a look at this, it would require the [Line 369](https://github.com/pinterest/pymemcache/blob/master/pymemcache/client/hash.py#L369) to be changed to `collections.defaultdict(collections.OrderedDict)`.
In this case the issue is caused by the `dict` class only being ordered from 3.6 and up (guaranteed from 3.7).
By switching to `OrderedDict` all versions will slow down, in order to satisfy a failing test for `py27, py34 and py35`.

I changed the benchmark to use the `HashClient` instead of the regular one, and got the following results from this command:
```sh
pytest --verbose --capture=no --no-cov -m benchmark pymemcache/test --keys 100 --count 10000 -k set_multi
```
- With `dict`: 41.0434 seconds.
- With `OrderedDict`: 41.7659 seconds.

Let me know what you think.

jparise commented:
> > > `test_set_many_unix` - It relies on the ordering of keys in a dict, which is unordered. I changed it from checking if a specific key failed, to just check a single key failed.
> > 
> > 
> > I think this would be a good place to use [`OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict). Otherwise, the test isn't particularly strong because it doesn't verify that the `NOT_STORED` item is the one that's actually returned.
> 
> I took a look at this, it would require the [Line 369](https://github.com/pinterest/pymemcache/blob/master/pymemcache/client/hash.py?rgh-link-date=2021-05-27T18%3A45%3A40Z#L369) to be changed to `collections.defaultdict(collections.OrderedDict)`.
> In this case the issue is caused by the `dict` class only being ordered from 3.6 and up (guaranteed from 3.7).
> By switching to `OrderedDict` all versions will slow down, in order to satisfy a failing test for `py27, py34 and py35`.

That makes sense. I originally thought the issue was only with the dictionary provided by the test function.

Thanks for the additional investigation!

martinnj commented:
@jparise - Sorry to nag a bit, is there any chance this can go into a small 3.4.4 release before the 4.0 milestone I see in the PR tab?
Have some legacy 2.7 code I'd like to have the change until it's ready to go full Python 3. :)

jparise commented:
> @jparise - Sorry to nag a bit, is there any chance this can go into a small 3.4.4 release before the 4.0 milestone I see in the PR tab?
> Have some legacy 2.7 code I'd like to have the change until it's ready to go full Python 3. :)

Yes, I'll publish a release shortly (#322).
